### PR TITLE
knex capabilities in blockchain search index - Closes #210

### LINKED ISSUE #210: Test Knex.js capabilities in blockchain search index
### Description

Similarly to #207, consider engaging [Knex](http://knexjs.org/) library in the blockchain search index.

### Motivation

- Need to know the technical limitations of storing blockchain data index

### Acceptance Criteria

- Comments in the issue with results
- Draft implementation for SQlite

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #210 

### How was it solved?
- [x]  Draft implementation for SQlite / MySQL
- [x]  Implement indexing and searches
- [x]  Performance comparision

### How was it tested?
Local + Jenkins

### CODE REVIEW

File: services/core/shared/database/knex.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+let dbConnection = null;
+
+const getDbInstance = async (tableName) => {
+    if (!dbConnection) dbConnection = await createDb();
+
+    const knex = dbConnection;
+
+    const write = async (row) => knex.transaction(async trx => {
+        const inserts = await trx(tableName).insert(row).onConflict('id').merge()
```
Comment (MichalTuleja) [inline services/core/shared/database/knex.js:56]: This needs a discussion - if the `write()` method overwrites data then there is a need for `update()`.
Now `write()` does updates by default.

File: services/core/shared/database/knex.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        } catch (error) {
+            const errCode = error.code;
+            const errMessage = error.message.split(`${errCode}: `)[1] || error.message;
+            const errCause = errMessage.split(': ')[0];
+            logger.debug('Encountered error with batch insert:', errCause,
+                '\nRe-attempting to update/merge the conflicted transactions one at a time: ');
+
+            return knex.transaction(async trx => {
+                const inserts = await Promise.all(rows.map(row => trx(tableName).insert(row).onConflict('id').merge()
+                    .transacting(trx)));
```
Comment (MichalTuleja) [inline services/core/shared/database/knex.js:79]: Same here, `write()` does updates.

File: services/core/shared/database/knex.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    const findAll = async () => knex.select().table(tableName);
+
+    const find = async (params) => {
+        const whereParams = params.selector;
+        const res = await knex.select().table(tableName).where(whereParams);
+        return res;
+    };
+
+    const findById = async (id) => {
+        const params = { selector: { id } };
```
Comment (MichalTuleja) [inline services/core/shared/database/knex.js:95]: Remove `selector` as it is specific to PouchDB.

**Review by MichalTuleja — COMMENTED** [review]
We need to discuss and resolve those issues before merging.

File: services/core/shared/database/knex.js (hunk 3 above)
Comment (nagdahimanshu) [reply services/core/shared/database/knex.js]: Done!

File: services/core/db_benchmarking/database/knex.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const createDb = async () => {
```
[the remark this answers is not in the document]
Comment (nagdahimanshu) [reply services/core/db_benchmarking/database/knex.js]: Done!

File: services/core/shared/database/knex.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    await knex.migrate.latest();
+
+    return knex;
+};
+
+let dbConnection = null;
+
+const getDbInstance = async (tableName) => {
+    if (!dbConnection) dbConnection = await createDb();
```
[the remark this answers is not in the document]
Comment (nagdahimanshu) [reply services/core/shared/database/knex.js]: Done

File: services/core/jobs/knex_migration.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+const config = require('../config');
+const getDbInstance = require('../shared/database/knex');
+
+const logger = Logger();
+
+module.exports = [
+	{
+		name: 'init.database.migration',
+		description: 'Initiate database migration',
+		schedule: '0 0 1 1 *', // Once a year
```
[the remark this answers is not in the document]
Comment (MichalTuleja) [reply services/core/jobs/knex_migration.js:27]: https://github.com/LiskHQ/lisk-service/issues/295

File: services/core/shared/database/knex.js (hunk 1 above)
Comment (MichalTuleja) [reply services/core/shared/database/knex.js:57]: #296

File: services/core/shared/database/knex.js (hunk 2 above)
Comment (MichalTuleja) [reply services/core/shared/database/knex.js:80]: #296

**Review by MichalTuleja — APPROVED** [review]

**Pull request merged**